src/main/scala/mesosphere/marathon/experimental/storage/BucketingPersistenceStore.scala

```diff
@@ -0,0 +1,178 @@
+    */
+  def buckets: Map[Int, String] = (0 to MaxBuckets).map(b => b -> f"$b%02d")(collection.breakOut)
+
+  /**
+    * Return bucket name calculated from the passed bucket hash.
+    */
+  def bucket(bucketNode: BucketNode): String = bucket(bucketNode.bucketPath)
+  def bucket(bucketPath: BucketPath): String = buckets(bucketPath.bucketHash % buckets.size)
+
+  /**
+    * Return full path for the passed [[BucketNode]]/[[BucketPath]] combining base path, bucket name and relative path.
+    */
+  def path(bucketNode: BucketNode): String = path(bucketNode.bucketPath)
+  def path(bucketPath: BucketPath): String = Paths.get(base, bucket(bucketPath), bucketPath.relativePath).toString
+
+  /**
+    * Helper stages, mapping from BucketNode/BucketPath to Node/Path
+    */
+  def toNode: Flow[BucketNode, Node, NotUsed] = Flow[BucketNode].map(bn => Node(path(bn), bn.payload))
+  def toPath: Flow[BucketPath, String, NotUsed] = Flow[BucketPath].map(bp => path(bp))
+
+  // All methods below wrap the storage methods by mapping from BucketNode/BucketPath to Node/Path
+  def create(bucketNode: BucketNode): Future[String] = store.create(Node(path(bucketNode), bucketNode.payload))
+  def create(bucketNodes: Seq[BucketNode]): Source[String, NotUsed] = Source(bucketNodes).via(createFlow)
+  def createFlow: Flow[BucketNode, String, NotUsed] =
+    Flow[BucketNode]
+      .via(toNode)
+      .via(store.createFlow)
+
+  def read(bucketPath: BucketPath): Future[Try[Node]] = store.read(path(bucketPath))
+  def read(bucketPaths: Seq[BucketPath]): Source[Try[Node], NotUsed] = Source(bucketPaths).via(readFlow)
+  def readFlow: Flow[BucketPath, Try[Node], NotUsed] =
+    Flow[BucketPath]
+      .via(toPath)
+      .via(store.readFlow)
+
+  def update(bucketNode: BucketNode): Future[String] = store.update(Node(path(bucketNode), bucketNode.payload))
+  def update(bucketNodes: Seq[BucketNode]): Source[String, NotUsed] = Source(bucketNodes).via(updateFlow)
+  def updateFlow: Flow[BucketNode, String, NotUsed] =
+    Flow[BucketNode]
+      .via(toNode)
+      .via(store.updateFlow)
+
+  def delete(bucketPath: BucketPath): Future[String] = store.delete(path(bucketPath))
+  def delete(bucketPaths: Seq[BucketPath]): Source[String, NotUsed] = Source(bucketPaths).via(deleteFlow)
+  def deleteFlow: Flow[BucketPath, String, NotUsed] =
+    Flow[BucketPath]
+      .via(toPath)
+      .via(store.deleteFlow)
+
+  def children(bucketPath: BucketPath): Future[Seq[String]] = store.children(path(bucketPath), true)
+  def children(bucketPaths: Seq[BucketPath]): Source[Seq[String], NotUsed] = Source(bucketPaths).via(childrenFlow)
+  def childrenFlow: Flow[BucketPath, Seq[String], NotUsed] =
+    Flow[BucketPath]
+      .via(toPath)
+      .via(store.childrenFlow(true))
+
+  def exists(bucketPath: BucketPath): Future[Boolean] = store.exists(path(bucketPath))
+  def exists(bucketPaths: Seq[BucketPath]): Source[Boolean, NotUsed] = Source(bucketPaths).via(existsFlow)
```

Comment:
Don't you think that `Source[Boolean, NotUsed]` can be replaced with `Future[Seq[Boolean]]`?

Reply:
I've changed the return type to a `(path: String, exists: Boolean)` to make it easier to know which path exists.